test/dynamo/test_ctx_manager.py

```diff
@@ -1644,6 +1730,1071 @@ def f(x):
+                x = torch.tensor([3.0, 3.0 + 5.0j])
+            return x
+
+        y = fn()
+        self.assertEqual(y.dtype, torch.complex128)
+        graph = eager.graphs[0]
+        actual = normalize_gm(graph.print_readable(False))
+
+        self.assertExpectedInline(
+            actual,
+            """\
+class GraphModule(torch.nn.Module):
+    def forward(self):
+        set_default_dtype = torch.set_default_dtype(torch.float64);  set_default_dtype = None
+
+        x: "c128[2]" = torch.tensor([3.0, (3+5j)])
+
+        set_default_dtype_1 = torch.set_default_dtype(torch.float32);  set_default_dtype_1 = None
+        return (x,)
+""",
+        )
+
+    def test_ctx_basic1(self):
+        @contextlib.contextmanager
+        def compute_sin(x):
+            try:
+                yield x.sin()
+            finally:
+                pass
+
+        @torch.compile(backend="eager", fullgraph=True)
+        def fn(x):
+            with compute_sin(x) as y:
+                return y.cos()
+
+        x = torch.tensor([1.0])
+        y = fn(x)
+        self.assertEqual(y, x.sin().cos())
+
+    def test_change_parent_nonlocal_0(self):
+        # test if a nonlocal actually gets propagated
+        z = 0
+        k = 0
+
+        def create_ctx():
+            @contextmanager
+            def ctx(x):
+                nonlocal z
+                nonlocal k
+                try:
+                    k = 100
+                    yield x.sin()
+                finally:
+                    pass
+
+            return ctx
+
+        def run_ctx(ctx, x):
+            nonlocal z
+            with ctx(x) as y:
+                z = k
+                return y.cos()
+
+        @torch.compile(backend="eager", fullgraph=True)
+        def fn(x):
+            ctx = create_ctx()
+            return run_ctx(ctx, x)
+
+        x = torch.tensor([1.0])
+        y = fn(x)
+        self.assertEqual(y, x.sin().cos())
+        self.assertEqual(z, 100)
+        self.assertEqual(k, 100)
+
+    def test_change_parent_nonlocal_1(self):
+        # test if finally is executed and it is reading the correct variable
+        z = 1
+        k = 2
+
+        def create_ctx():
+            @contextmanager
+            def ctx(x):
+                nonlocal z
+                nonlocal k
+                try:
+                    yield x.sin()
+                finally:
+                    k = z
+
+            return ctx
+
+        def run_ctx(ctx, x):
+            nonlocal z
+            z = 100
+            with ctx(x) as y:
+                return y.cos()
+
+        @torch.compile(backend="eager", fullgraph=True)
+        def fn(x):
+            ctx = create_ctx()
+            return run_ctx(ctx, x)
+
+        x = torch.tensor([1.0])
+        y = fn(x)
+        self.assertEqual(y, x.sin().cos())
+        self.assertEqual(z, 100)
+        self.assertEqual(k, 100)
+
+    @unittest.expectedFailure
+    def test_change_parent_global_0(self):
```

Comment:
What's the explanation for this one?

Reply:
Sorry. This one slipped out of my hands. Will take a closer look at this one.

Reply:
Would it be ok if I handle this one in a follow-up PR?

Reply:
It depends on the severity of the problem. If we need a design change then it should happen in this PR :P

Also, while the total LOC in this PR looks long, it's actually not that bad, most of the changes are in the tests.

Reply:
Fair enough.

Reply:
Ok. I think I know what's wrong in this case. In the example below, `foo` and `bar` have different `__globals__` and the `STORE_GLOBAL` instruction will not propagate the assignment to the parent frame.

```python
In [10]: def foo():
    ...:     @contextlib.contextmanager
    ...:     def bar():
    ...:         try:
    ...:             yield z
    ...:         finally:
    ...:             pass
    ...:     return bar
    ...:

In [12]: id(foo().__globals__)
Out[12]: 4319810112

In [13]: id(foo.__globals__)
Out[13]: 4355061184
```

It should go into the first `if` stmt but executes the `else` branch.
https://github.com/pytorch/pytorch/blob/82ce8882730791e202246f431d6495d1efd4f7d7/torch/_dynamo/symbolic_convert.py#L3329-L3333

Reply:
I was able to create a reproducer for this bug **without** context manager. I'll create an issue for it and mark the test as `@expectedFailure`